File: frontend/lib/types/generation.ts
export interface GenerationJob {
  id: string;
  project_id: string;
  status: "PENDING" | "RUNNING" | "COMPLETE" | "FAILED" | "PAUSED" | "BUDGET_EXHAUSTED";
  total_shots: number | null;
  completed_shots: number;
  estimated_cost: number | null;
  actual_cost: number;
  created_at: string;
  completed_at: string | null;
}

/** One bible reference that conditioned a clip's generation. */
export interface ClipReference {
  url: string;
  role: "identity" | "costume" | "prev_frame" | "location" | "style";
  character?: string;
}

export interface GeneratedClip {
  id: string;
  job_id: string;
  shot_id: string;
  model_used: string | null;
  prompt: string | null;
  url: string | null;
  consistency_score: number | null;
  references_json?: ClipReference[] | null;
  seed?: number | null;
  cost_usd?: number | null;
  status: ClipStatus;
  retries: number;
  created_at: string;
}

export type ClipStatus = "PENDING" | "GENERATING" | "CHECKING" | "APPROVED" | "FAILED" | "NEEDS_REVIEW" | "TIMEOUT" | "SKIPPED";

File: backend/app/schemas/generation.py
from pydantic import BaseModel
from uuid import UUID
from typing import Optional


class GenerationStartRequest(BaseModel):
    project_id: UUID


class GenerationJobStatus(BaseModel):
    id: UUID
    status: str
    total_shots: Optional[int] = None
    completed_shots: int = 0
    estimated_cost: Optional[float] = None
    actual_cost: float = 0.0

    model_config = {"from_attributes": True}


class ClipResult(BaseModel):
    id: UUID
    shot_id: UUID
    model_used: Optional[str] = None
    url: Optional[str] = None
    consistency_score: Optional[float] = None
    # bible references that conditioned this clip ([{url, role, character?}])
    references_json: Optional[list] = None
    seed: Optional[int] = None
    status: str
    retries: int = 0
    cost_usd: Optional[float] = None
    # the clip's REAL probed duration (models render short of the request)
    duration_seconds: Optional[float] = None

    model_config = {"from_attributes": True}
